services/py-api/src/database/transaction_manager.py

```diff
@@ -0,0 +1,87 @@
+# mypy: disable-error-code=no-any-return
+# This is because we have Generic methods
+
+from time import monotonic
+from typing import Callable, Any, TypeVar, Awaitable
+
+from pymongo.errors import PyMongoError
+from result import Err, is_err, Result
+from structlog.stdlib import get_logger
+
+from src.database.db_manager import DatabaseManager
+
+LOG = get_logger()
+
+T = TypeVar("T")
+E = TypeVar("E")
+
+
+class TransactionManager:
+    _RETRY_TRANSACTION_DEADLINE_SEC = 3
+
+    def __init__(self, db_manager: DatabaseManager) -> None:
+        self._client = db_manager.client
+
+    def _deadline_exceeded(self, start_time: float) -> bool:
+        return monotonic() - start_time > self._RETRY_TRANSACTION_DEADLINE_SEC
+
+    async def _retry_tx(self, func: Callable[..., Awaitable[Result[T, E]]], *args: Any, **kwargs: Any) -> Result[T, E]:
```

Comment:
I would suggest that you write a good documentation around this class as it might be pretty confusing to a newcomer especially the types.

Reply:
Fixed in 3dbcb9b202de882ccd1fe0025df48c3c6fff976c